### PR TITLE
[PIP 86][Function] add funtion api to support preload and release external resources

### PR DESCRIPTION
**Pull request by wangjialing218**
### Motivation

This patch is based on #11112, I have talked with @wuchenxi123 and I will continue the work for this feature.

### Modifications

Introduce `RichFunction` interface extends `Function`,  which provide `setup` and `tearDown` API.
`setup` is called only once when function instance started. `tearDown` is called when function instance closed.
User could use these interface to initialize and release their external resources such as RedisClient.

### Does this pull request potentially affect one of the following parts:
*If `yes` was chosen, please highlight the changes*
  - The public API: ( **yes**)

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 

- [x] `doc-required` 
I think we need to add chapter to tell user how to use the API in `functions-develop.md`

### CODE REVIEW

**Comment by nlu90** [comment]
@wangjialing218 Did you call a vote for the PIP-86 in the dev@pulsar mailing list?

It's good if we can get feedback from the community. And then start the review work.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java [hunk 1]
```diff
@@ -229,7 +229,9 @@ synchronized private void setup() throws Exception {
         setupLogHandler();
 
         javaInstance = new JavaInstance(contextImpl, object, instanceConfig);
-
+        Thread.currentThread().setContextClassLoader(functionClassLoader);
+        javaInstance.setup();
+        Thread.currentThread().setContextClassLoader(instanceClassLoader);
```
Comment (eolivelli) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java:234]: We should do this I  a finally block

**Review by eolivelli — CHANGES_REQUESTED** [review]
Overall is good.
Please add unit tests and integration tests.

There is a duplicate PR for this PIP, please close the old one

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java (hunk 1 above)
Comment (wangjialing218) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java:234]: Sorry I did not catch this.
If exception happened in user's setup() code,  `JavaInstanceRunnable.run()` will catch the exception and close function instance. Why we need to do this in a finally block

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/RichFunction.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.apache.pulsar.common.classification.InterfaceAudience;
+
+/**
+ * An interface for richFunction
+ * Extension for Function interface
+ * Provide setup and tearDown interface to initial and close any resource used in function
+ */
+@InterfaceAudience.Public
+public interface RichFunction extends Function{
```
Comment (niclash) [inline pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/RichFunction.java:29]: I am not happy with the API naming. `InitializableFunction` is far more precise name. And `setup()` and `teardown()` sound like unit testing. Suggestion `initialize()` and `close()`.

Comment (wangjialing218) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/RichFunction.java:29]: Sounds good. I will change the naming.

Comment (merlimat) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/RichFunction.java]: Actually, I'd suggest to instead add these methods in the `org.apache.pulsar.functions.api.Function`, with default implementations that doesn't do anything.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java (hunk 1 above)
Comment (merlimat) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java:234]: We need to make to reset the class loader to the `instanceClassLoader`.

Comment (mattisonchao) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java:234]: > We need to make sure to reset the class loader to the `instanceClassLoader`.

+1

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/RichFunction.java (hunk 2 above)
Comment (niclash) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/RichFunction.java]: @merlimat Even better!

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Function.java [hunk 3]
```diff
@@ -38,4 +38,20 @@
      * @return the output
      */
     O process(I input, Context context) throws Exception;
+
+    /**
+     * Called when function instance start
```
Comment (nlu90) [inline pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Function.java:43]: `Called once to initialize resources when function instance is started.`

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Function.java [hunk 4]
```diff
@@ -38,4 +38,20 @@
...
+     *
+     * @param context The Function context
+     *
+     * @throws Exception
+     */
+    default void initialize(Context context) throws Exception {}
+
+    /**
+     * Called when function instance close
```
Comment (nlu90) [inline pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Function.java:52]: Called once to properly close resources when function instance is stopped.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java [hunk 5]
```diff
@@ -155,10 +155,24 @@ private void processAsyncResults(JavaInstanceRunnable.AsyncResultConsumer result
...
+            function.initialize(context);
+        }
+    }
+
     @Override
     public void close() {
         context.close();
         executor.shutdown();
+
+        if (function != null) {
```
Comment (nlu90) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java:169]: I think we should first close the function resources before close context and executor.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java [hunk 6]
```diff
@@ -453,7 +458,9 @@ synchronized public void close() {
         }
 
         if (null != javaInstance) {
+            Thread.currentThread().setContextClassLoader(functionClassLoader);
             javaInstance.close();
+            Thread.currentThread().setContextClassLoader(instanceClassLoader);
```
Comment (nlu90) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java:463]: need `try ... finally ...`  to make sure the classloader is reset correctly

**Review by nlu90 — COMMENTED** [review]
The PIP Vote has been passed.

And some updates in the PIP-86 needed to reflect the actual changes we added here.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java (hunk 1 above)
Comment (wangjialing218) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java]: done

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/RichFunction.java (hunk 2 above)
Comment (wangjialing218) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/RichFunction.java]: moved methods to `Funtion`

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java (hunk 6 above)
Comment (wangjialing218) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java]: done

**Comment by wangjialing218** [comment]
@nlu90 @eolivelli please have a check again

**Review by nlu90 — APPROVED** [review]
LGTM

**Review by merlimat — APPROVED** [review]

**Review by eolivelli — CHANGES_REQUESTED** [review]
The change looks good.
But in your test your are not testing that the new methods are executed.
I suggest to add some static variables in the function in order to be able to check that the function runs

**Comment by wangjialing218** [comment]
> The change looks good. But in your test your are not testing that the new methods are executed. I suggest to add some static variables in the function in order to be able to check that the function runs

@eolivelli I tried to add static variables in function but failed to get correct value in test code, may be caused by different class loader, could you give some suggestion?
Anyway, in the test funtion, `process()` will throw exception if `initialize()` did not run, and test code could check whether all messages received.

**Review by freeznet — APPROVED** [review]

**Comment by nlu90** [comment]
@eolivelli Could you help take a look again when you are available?

**Review by eolivelli — APPROVED** [review]
+1

**Pull request merged**